### PR TITLE
add declare_file with sibling as ArtifactRoot

### LINKED ISSUE #4858: not possible to declare_file to genfiles_dir
### Description of the problem / feature request:

since ctx.new_file() is deprecated ctx.actions.declare_file() should be a replacement.
with ctx.new_file() it was possible to generate files to genfiles_dir which I tried with ctx.actions.declare_file()

```python
out = ctx.actions.declare_file(ctx.attr.out, sibling=ctx.genfiles_dir)
```

produces

```
Cannot convert parameter 'sibling' to type File or NoneType, in method declare_file(string, root sibling) of 'actions'
```

### Feature requests: what underlying problem are you trying to solve with this feature?

it must be possible to use a root dir (ctx.bin_dir and ctx.genfiles_dir) as sibling
or as different parameter e.g. root.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

rule.bzl:
```python
def _impl(ctx):
    o = ctx.actions.declare_file('x.sh', sibling=ctx.bin_dir)
    ctx.actions.write(o, 'echo Stuff', is_executable = True)
    return [DefaultInfo(executable = o)]

my_rule = rule(_impl, executable = True)
```

BUILD:
```python
load(':rule.bzl', 'my_rule')
my_rule(name = 'xxx')
```
`bazel build xxx`

### What operating system are you running Bazel on?

macOS 10.12.6

### What's the output of `bazel info release`?

release 0.11.1-homebrew

### PR DESCRIPTION
**Pull request by ahippler**
Fixes #4858

### CODE REVIEW

**Comment by ahippler** [comment]
is it cleaner to use a third parameter for the roots?

after the review I would also change declare_directory() and add a test for it.
there are no test at all for declare_directory().

File: src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java [hunk 1]
```diff
@@ -130,25 +130,31 @@ ArtifactRoot newFileRoot() throws EvalException {
         doc =
             "A file that lives in the same directory as the newly created file. "
                 + "The file must be in the current package.",
```
Comment (dslomov) [inline src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java]: Please update this doc string

File: src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java [hunk 2]
```diff
@@ -166,7 +172,10 @@ public Artifact declareFile(String filename, Object sibling) throws EvalExceptio
           @Param(
               name = "sibling",
               doc = "A file that lives in the same directory as the newly declared directory.",
```
Comment (dslomov) [inline src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java]: Please update this doc string

File: src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java [hunk 3]
```diff
@@ -129,61 +129,112 @@ ArtifactRoot newFileRoot() throws EvalException {
...
+    name = "declare_directory",
+    doc =
+        "Declares that rule or aspect create a directory with the given name, in the "
+            + "current package. You must create an action that generates the directory.",
+    parameters = {
+      @Param(
+        name = "filename",
+        type = String.class,
+        doc =
+            "If no 'sibling' provided, path of the new directory, relative "
```
Comment (dslomov) [inline src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java]: What about `root`? Same for `declare_file` above.

File: src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java [hunk 4]
```diff
@@ -129,61 +129,112 @@ ArtifactRoot newFileRoot() throws EvalException {
...
-  public Artifact declareDirectory(String filename, Object sibling) throws EvalException {
+  public Artifact declareDirectory(String filename, Object sibling, Object root) throws EvalException {
     context.checkMutable("actions.declare_directory");
-    if (Runtime.NONE.equals(sibling)) {
-      return ruleContext.getPackageRelativeTreeArtifact(
-          PathFragment.create(filename), newFileRoot());
-    } else {
+    if (sibling instanceof  Artifact && root instanceof ArtifactRoot) {
+      throw new EvalException(
+        ruleContext.getRule().getAttributeLocation("sibling"),
```
Comment (dslomov) [inline src/main/java/com/google/devtools/build/lib/analysis/skylark/SkylarkActionFactory.java]: This is wrong. It is a function call not a rule. You can just throw EvalException will null location, and the correct location should be attached to that. Validate that with a test.

**Comment by dslomov** [comment]
I apologize, the bug has got some pushback, as you can see. The code is good to go, but let's decide whether this is the right change in the first place.

Apologies again, I should have flagged this earlier.

**Comment by ahippler** [comment]
@dslomov is there a final decision?

**Comment by ulfjack** [comment]
Ping @dslomov

**Comment by jin** [comment]
ping @dslomov, what's the status of the merge here?

**Comment by meisterT** [comment]
genfiles have been removed in #6761, so this PR is obsolete

**Pull request closed without merging**